dbgd.c

```diff
@@ -298,3 +300,65 @@ static int dbgd_show_front_screen(Dbg__Request *req, Dbg__Response *res)
 
     return DBG__RESPONSE__TYPE__OK;
 }
+
+static int dbgd_call(Dbg__Request *req, Dbg__Response *res)
+{
+    if (!req->has_address)
+        return DBG__RESPONSE__TYPE__ERROR_INCOMPLETE_REQUEST;
+
+    // These variables will be used as parameters for inline assembly.
+    // We make them static as the stack pointer will be changed.
+    // So we want to address them globally.
+    static uint32_t stack_pointer;
+    static uint32_t stack_backup;
+    static uint32_t address;
+
+    // Allocate a stack for working and space for supplied data
```

Comment:
One thing to consider... We could potentially simplify this and just let the client explicitly allocate the stack page and pass in the appropriate ESP value (at the potential cost of latency, mitigated perhaps in the future with command batching). This would allow us to preserve the stack page across calls (giving us the ability to read any modified data structures) and simplify the code in this function considerably.

Reply:
Agreed, hence my note in the PR description about possible future API changes.

One issue I tried to solve here is the extra communication overhead, in particular when using XBDM instead of nxdk-rdt. XBDM only does a certain number of commands per second.

**Currently:**

1. CALL (Allocate stack, write stack, run code &rarr; eax is returned)

**If we split this:**

1. ALLOC (Allocate stack &rarr; address is returned)
2. MEM_WRITE (Write stack contents &rarr; nothing is returned)
3. CALL (Call &rarr; eax is returned)

This is also why the register set is immediately returned: it has large performance implications on XBDM (as we can avoid an additional readback command).
In xboxpy, the XBDM packet handler is also entirely written in assembly (making nicer, more complicated solutions harder to implement).

This also affects usability as simple calls suddenly become multi-step operations. The RDTSC example is probably complicated enough as-is for most users.